pkg/pub_worker/lib/src/bin/pana_wrapper.dart

```diff
@@ -197,61 +197,130 @@ Future<(SdkConfig, SdkConfig)> _detectSdks(Pubspec pubspec) async {
+    bool allowsMissingVersion = false,
+  }) {
+    if (!allowsMissingVersion && (dart == null || flutter == null)) {
+      return false;
+    }
 
-  var dartSdkPath = installedDartSdk?.path;
-  var flutterSdkPath = installedFlutterSdk?.path;
-  String configKind = 'stable';
-  if (needsNewer) {
-    configKind = 'preview';
-    dartSdkPath = await _previewDartSdk() ??
-        dartSdks.firstWhereOrNull((sdk) => sdk.version.isPreRelease)?.path ??
-        dartSdkPath;
-    flutterSdkPath = await _previewFlutterSdk() ??
-        flutterSdks.firstWhereOrNull((sdk) => sdk.version.isPreRelease)?.path ??
-        flutterSdkPath;
+    return sdkMatchesConstraint(
+          sdkVersion: dart,
+          constraint: pubspec.dartSdkConstraint,
+        ) &&
+        sdkMatchesConstraint(
+          sdkVersion: flutter,
+          constraint: pubspec.flutterSdkConstraint,
+        );
+  }
+
+  // try to use the latest SDKs in the docker image
+  final matchesInstallSdks = matchesSdks(
+    dart: installedDartSdk?.version,
+    flutter: installedFlutterSdk?.version,
+    allowsMissingVersion: true,
+  );
+  if (matchesInstallSdks) {
+    return (
+      SdkConfig(
+        rootPath: installedDartSdk?.path,
+        configHomePath: _configHomePath('dart', 'stable'),
+      ),
+      SdkConfig(
+        rootPath: installedFlutterSdk?.path,
+        configHomePath: _configHomePath('flutter', 'stable'),
+      ),
+    );
   }
 
+  final flutterArchive = await fetchFlutterArchive();
+
+  // try to use the latest stable downloadable SDKs
+  final latestStableDartSdk =
+      await fetchLatestDartSdkVersion(channel: 'stable');
+  final latestStableDartSdkVersion = latestStableDartSdk?.semanticVersion;
+  final matchesLatestStable = matchesSdks(
+    dart: latestStableDartSdkVersion,
+    flutter: flutterArchive?.latestStable?.semanticVersion,
+  );
+  if (matchesLatestStable) {
+    final dartSdkPath = await _installSdk(
+      sdkKind: 'dart',
+      configKind: 'latest-stable',
+      version: latestStableDartSdkVersion!.toString(),
+    );
+    final flutterSdkPath = await _installSdk(
+      sdkKind: 'flutter',
+      configKind: 'latest-stable',
+      version: flutterArchive!.latestStable!.version!,
+    );
+    return (
+      SdkConfig(
+        rootPath: dartSdkPath,
+        configHomePath: _configHomePath('dart', 'latest-stable'),
+      ),
+      SdkConfig(
+        rootPath: flutterSdkPath,
+        configHomePath: _configHomePath('flutter', 'latest-stable'),
+      ),
+    );
+  }
+
+  // try to use the latest beta/preview SDKs
+  final latestBetaDartSdk = await fetchLatestDartSdkVersion(channel: 'beta');
+  final latestBetaDartSdkVersion = latestBetaDartSdk?.semanticVersion;
+  final matchesLatestBeta = matchesSdks(
+    dart: latestBetaDartSdkVersion,
+    flutter: flutterArchive?.latestBeta?.semanticVersion,
+  );
+  if (matchesLatestBeta) {
+    final dartSdkPath = await _installSdk(
+      sdkKind: 'dart',
+      configKind: 'latest-beta',
```

Comment:
Would it be possible to ask for the specific version we think is the latest beta, instead of just asking for the "latest beta" (same for stable above) to avoid however small a race.

Reply:
Not sure: what is the race here?

Reply:
If you detect that latest stable is good
a new stable is released
you ask to install latest table

The version you checked against will not be the version you install

Reply:
Oh, the `latest-beta` is just the name for (a) the SKD directory as in `/home/worker/dart/latest-beta`, and also for its configuration directory as in `/home/worker/config/latest-beta`. The version we are installing (the `version` parameter) is the same we are checking the constraint against.

Reply:
Ah - then all is good. I thought you were passing the channel name instead of the version